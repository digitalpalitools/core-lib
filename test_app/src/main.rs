use rusqlite::{Connection, Row, NO_PARAMS};

fn get_row_cells(row: &Row) -> Vec<String> {
    let cells: Vec<String> = row
        .column_names()
        .iter()
        .map(|&cn| {
            let cell: String = match row.get(cn) {
                Ok(c) => c,
                Err(e) => e.to_string(),
            };
            cell
        })
        .collect();

    cells
}

fn exec_sql_core(sql: &str) -> rusqlite::Result<Vec<Vec<Vec<String>>>, rusqlite::Error> {
    let conn = Connection::open("./inflections.db")?;
    let mut result: Vec<Vec<Vec<String>>> = Vec::new();
    for s in sql.split(';').filter(|s| !s.trim().is_empty()) {
        let mut stmt = conn.prepare(&s)?;
        let mut rows = stmt.query(NO_PARAMS)?;

        let mut table: Vec<Vec<String>> = Vec::new();
        while let Some(row) = rows.next()? {
            table.push(get_row_cells(row));
        }
        result.push(table)
    }

    Ok(result)
}

fn exec_sql(sql: &str) -> Result<String, String> {
    let table = exec_sql_core(&sql).map_err(|x| x.to_string())?;
    serde_json::to_string(&table).map_err(|x| x.to_string())
}

fn main() -> Result<(), String> {
    println!("{:?}", pls_core::alphabet::PALI_ALPHABET_ROMAN);
    let x = pls_core::alphabet::PaliAlphabet::AA;
    println!("ā > bh? {:#?}", x > pls_core::alphabet::PaliAlphabet::BH);

    let html = pls_core::inflections::generate_inflection_table(
        "vyābādhetvā",
        |s| Ok(s.to_string()),
        exec_sql,
    )?;
    println!("{:#?}", html);

    std::fs::write("d:/delme/inflections.txt", &html).unwrap();

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    // TODO: Need to be abstracted, granularized and moved to pls_core.
    fn basic_inflection_test() {
        let html = pls_core::inflections::generate_inflection_table(
            "ābādheti",
            |s| Ok(s.to_string()),
            exec_sql,
        );

        let approved_html = include_str!("test_data/basic_inflection_test.approved.txt");

        assert_eq!(html.unwrap(), approved_html);
    }
}
