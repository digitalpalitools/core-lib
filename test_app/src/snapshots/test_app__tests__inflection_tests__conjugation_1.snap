---
source: test_app/src/main.rs
expression: html

---
<div class="pls-inflection-root">
  <header class="pls-inflection-header"><summary class="pls-inflection-summary">ābādheti &ndash; "eti pr" (like bhāveti)</summary></header>
<table class="pls-inflection-table pls-inflection-type-conjugation">
    <thead>
      <tr>
        <td>pr</td>
        <td colspan="2">active</td>
        <td colspan="2">reflexive</td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>person</td>
        <td>singular</td>
        <td>plural</td>
        <td>singular</td>
        <td>plural</td>
      </tr>
      <tr>
        <td>pr 3rd</td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">eti</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">enti</span></div>
          </td>
        <td>
          </td>
        <td>
          </td>
      </tr>
      <tr>
        <td>pr 2nd</td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">esi</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">etha</span></div>
          </td>
        <td>
          </td>
        <td>
          </td>
      </tr>
      <tr>
        <td>pr 1st</td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">emi</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">ema</span></div>
          </td>
        <td>
          </td>
        <td>
          </td>
      </tr>
    </tbody>
  </table>

<table class="pls-inflection-table pls-inflection-type-conjugation">
    <thead>
      <tr>
        <td>imp</td>
        <td colspan="2">active</td>
        <td colspan="2">reflexive</td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>person</td>
        <td>singular</td>
        <td>plural</td>
        <td>singular</td>
        <td>plural</td>
      </tr>
      <tr>
        <td>imp 3rd</td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">etu</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">entu</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">etaṃ</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">entaṃ</span></div>
          </td>
      </tr>
      <tr>
        <td>imp 2nd</td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">ehi</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">etha</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">essu</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">evho</span></div>
          </td>
      </tr>
      <tr>
        <td>imp 1st</td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">emi</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">ema</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">e</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">emase</span></div>
          </td>
      </tr>
    </tbody>
  </table>

<table class="pls-inflection-table pls-inflection-type-conjugation">
    <thead>
      <tr>
        <td>opt</td>
        <td colspan="2">active</td>
        <td colspan="2">reflexive</td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>person</td>
        <td>singular</td>
        <td>plural</td>
        <td>singular</td>
        <td>plural</td>
      </tr>
      <tr>
        <td>opt 3rd</td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">eyya</span></div>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">e</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">eyyuṃ</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">etha</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">eraṃ</span></div>
          </td>
      </tr>
      <tr>
        <td>opt 2nd</td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">eyyāsi</span></div>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">e</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">eyyātha</span></div>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">etha</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">etho</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">eyyavho</span></div>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">eyyāvho</span></div>
          </td>
      </tr>
      <tr>
        <td>opt 1st</td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">eyyāmi</span></div>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">eyyaṃ</span></div>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">e</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">eyyāma</span></div>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">ema</span></div>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">emu</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">eyyaṃ</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">eyyāmhe</span></div>
          </td>
      </tr>
    </tbody>
  </table>

<table class="pls-inflection-table pls-inflection-type-conjugation">
    <thead>
      <tr>
        <td>fut</td>
        <td colspan="2">active</td>
        <td colspan="2">reflexive</td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>person</td>
        <td>singular</td>
        <td>plural</td>
        <td>singular</td>
        <td>plural</td>
      </tr>
      <tr>
        <td>fut 3rd</td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">essati</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">essanti</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">essate</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">essante</span></div>
          </td>
      </tr>
      <tr>
        <td>fut 2nd</td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">essasi</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">essatha</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">essase</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">essavhe</span></div>
          </td>
      </tr>
      <tr>
        <td>fut 1st</td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">essāmi</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">essāma</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">essaṃ</span></div>
          </td>
        <td>
          <div class="pls-inflection-inflected-word">ābādh<span class="pls-inflection-inflected-word-suffix">essāmhe</span></div>
          </td>
      </tr>
    </tbody>
  </table>


  <footer class="pls-inflection-footer"><a class="pls-inflection-feedback-link" target="_blank" href="https://docs.google.com/forms/d/e/1FAIpQLSdqnYM0_5VeWzkFBPzyxaLqUfKWgNjI8STCpdrx4vX3hetyxw/viewform"><strong>spot a mistake? something missing? fix it here!</strong></a></footer>
</div>

